source/server/server.cc

```diff
@@ -463,5 +463,17 @@ ProtobufTypes::MessagePtr InstanceImpl::dumpBootstrapConfig() {
   return config_dump;
 }
 
+void InstanceImpl::adminRequest(absl::string_view path, const Http::Utility::QueryParams& params,
+                                absl::string_view method, AdminResponseHandler& response_handler) {
+  std::string path_copy = std::string(path);
+  std::string method_copy = std::string(method);
+  dispatcher_->post([this, path_copy, params, method_copy, response_handler]() {
+    Http::HeaderMapImpl response_headers;
+    std::string body;
+    Http::Code code = admin_->request(path_copy, params, method_copy, response_headers, body);
+    response_handler(code, response_headers, body);
```

Comment:
Are you sure this is the pattern we want here? The callback is going to execute on some other thread (i.e. the main thread) to that of the caller. So, there will need to be some cross-thread synchronization happening. This has been a total nightmare in the `FakeUpstream` integration test code, and might lead to problematic main thread blocking.

An alternative would be to have the calling thread allocate a buffer, pass it to the `post`, wait using a `CondVar` for a response, and then take the buffer. That works nicely if the calling thread is not on an event loop. Alternatively, you post the response buffer to the event loop of the calling thread when done if async.

Reply:
Why isn't this always just happening on the main thread? Do we really need to support post here at all?

Reply:
The caller will likely come from another thread, at least in our case.

Reply:
WDYT about handling the threading complexity in our private code then? I think for most people a use of this would be on the main thread...

Reply:
It sounds reasonable to have this PR just define Admin::request() requiring it be called from the main thread, so there's no complexity in this PR.

And I realize now that sever()->dispatcher()->post() provides all the infrastructure I'd need to call this from a different thread. I could make Admin::request() call DispatcherImpl::isThreadSafe() I suppose, if that were public, but if that's not generally going to happen I suppose it's sufficient to just document it.

I'm a little reluctant to do the post/block in this PR because that would tie up the calling thread, when the system we are merging it into might want to continue to do other work on that thread while the post is being processed.